File: src/309.best_time_to_buy_and_sell_stock_with_cooldown.kt
fun main() {
    println(maxProfit4(intArrayOf(1, 2, 3, 0, 2)))
    println(maxProfit4(intArrayOf(1)))
}

fun maxProfit4(prices: IntArray): Int {
    var sold = -prices.first()
    var held = -prices.first()
    var rest = 0

    for (i in 1 ..< prices.size) {
        val prvSold = sold
        sold = held + prices[i]
        held = Math.max(held, rest - prices[i])
        rest = Math.max(rest, prvSold)
    }

    return Math.max(rest, sold)
}

File: lib/309.best_time_to_buy_and_sell_stock_with_cooldown.dart
import 'dart:math';

void main() {
  print(maxProfit([1,2,3,0,2]));
  print(maxProfit([1]));
}

/// todo
int maxProfit(List<int> prices) {
  int sold = -prices.first;
  int held = -prices.first;
  int rest = 0;

  for(int i = 1; i < prices.length; i ++) {
    int prvSold = sold;
    sold = held + prices[i];
    // rest or buy p[i]
    held = max(held, rest - prices[i]);
    // rest again or rest once and accept the prv sold value
    rest = max(rest, prvSold);
  }

  return max(rest, sold);
}
